### PR TITLE
Make scan batch get identifiers en masse for children as opposed to o…

### PR DESCRIPTION
**Pull request by jordandukart**
…ne at a time.

**JIRA Ticket**: https://jira.duraspace.org/browse/ISLANDORA-1855

# What does this Pull Request do?

When batches that extend scan batch (Islandora Book Batch / Islandora Newspaper Batch for example) go to preprocess children identifiers from Fedora will be requested all children at once in one request as opposed to doing a separate request for each child (n requests).

# What's new?
Requests are reduced from n to 1 for children of an object. This has big performance implications for objects that have a lot of direct children such as book pages.

# How should this be tested?

Use the Islandora Book Batch from the command line or the UI to ingest a book. Note that the `getNextPID` calls for pages ask for n pids at once as opposed to one at a time in the Fedora logs.

Example for two books with two pages each:
```
INFO 2017-02-09 17:45:42.417 [http-8080-10] (DefaultManagement) Completed getNextPID(numPIDs: 1, namespace: islandora)
INFO 2017-02-09 17:45:42.423 [http-8080-10] (DefaultManagement) Completed getNextPID(numPIDs: 2, namespace: islandora)
INFO 2017-02-09 17:45:42.435 [http-8080-10] (DefaultManagement) Completed getNextPID(numPIDs: 1, namespace: islandora)
INFO 2017-02-09 17:45:42.440 [http-8080-10] (DefaultManagement) Completed getNextPID(numPIDs: 2, namespace: islandora)

```

# Interested parties
@DiegoPino, @adam-vessey, @Islandora/7-x-1-x-committers

### CODE REVIEW

**Comment by DiegoPino** [comment]
@jordandukart looks fine in terms of performance, have to read the new logic(namespace thing, etc), the only issue on a first quick code view is that if you request 1000 PIDS for a book with 1000 pages at once, and something fails in between (like the second page dies, or the parent, since there is this preprocess of children right), well, then you lost 1000 PIDs. with UUID not so bad, with integer incremental PIDs kinda bad. Just a thought.

**Comment by jordandukart** [comment]
I feel like this is an acceptable trade off and if Fedora PIDs are being used for identifier mapping UUIDs should likely be enforced imo.

**Comment by jordandukart** [comment]
@DiegoPino,

Namespace logic comes from the preprocessor base: https://github.com/Islandora/islandora_batch/blob/7.x/includes/preprocessor_base.inc#L116-L126, saves you some diggin'.

**Comment by DiegoPino** [comment]
@jordandukart I do feel it's an acceptable tradeoff, true. Just need some time to test. Also, wondering (just now, came to my mind) if this particular logic should not be better implemented in those classes that extend scan batch? It only affects really books, newspapers. right? Just wondering... just wondering

**Comment by pc37utn** [comment]
knowing how pids are generated in certain ingests is important to some of our workflows.   I would rather cloudsync "book1:*"  than whatever you would have to do to find the UUIDs for a certain volume.

**Comment by adam-vessey** [comment]
@pc37utn: I may be misinterpreting, but it appears that you're selecting everything in a given namespace. This would still be possible with UUID PIDs as they're presently implemented: Namespaces are still used with the UUID, the UUID just replaces the numerical portion of the PID.

**Comment by willtp87** [comment]
@DiegoPino  I think it is good to do this in scan batch because in a practical (and documented in code) sense it acts as a set of base classes.  Most batches derive from it and this will let them benefit from the new shared code much like they currently do from the existing shared code.

**Comment by jordandukart** [comment]
I am onboard for whatever approach, IIRC committers decided that it would make sense to do this at the scan batch level so that the performance gains would roll down to the existing batches. If we think this is no longer the best course of action I can create a new preprocess class that extends scan and overrides the one function so people can implement it as they see fit.

**Comment by DiegoPino** [comment]
@jordandukart @willtp87 yeah makes sense as it is. Probably overthinking this. Will review during the day, have some time on the train. Thanks!

**Comment by willtp87** [comment]
I gave this a test and saw no issue.

File: includes/islandora_scan_batch.inc
Comment (DiegoPino) [inline includes/islandora_scan_batch.inc]: Maybe i'm a bit overconcern with this, so it's just a comment: Imports are a sensitive topic and when things get wrong they get wrong Big. Would it not be better to iterate over $children instead of $children_ids? children_ids is hitting directly the repository and I can see edge cases (unprovable but still hard to debug) around that could make ` $this->connection->repository->getNextIdentifier()` respond weirdly which would make objects (children) being silently dropped. If the iteration is over the children itself, then on missing PID's for a certain child, there could be a fallback or even an exception. Does that makes any sense?

Comment (jordandukart) [reply includes/islandora_scan_batch.inc]: Can you elaborate @DiegoPino?

**Comment by jordandukart** [comment]
Also @adam-vessey noticed that when one PID is requested Tuque returns a string not an array, so casting to an array: https://github.com/Islandora/tuque/blob/ac7ab6a2f577d59e48f3e1f899bba82cd3e2a44e/FedoraApi.php#L938.

File: includes/islandora_scan_batch.inc
Comment (DiegoPino) [reply includes/islandora_scan_batch.inc]: You are iterating over the Identifiers you get on line 201. Not on the original to-be-created-objects-array named $children. Even when the difference should not exist (means both arrays should have the same number of items) in the case of any failure/weird thing/tuque-failing on provide the PIDSs, it does not seem to be an approach I would encourage.

**Review by DiegoPino — APPROVED** [review]

**Comment by DiegoPino** [comment]
@jordandukart appreciate the change on that iteration. @adam-vessey and @willtp87 are you all OK with this? I feel doing something more than an exception on not allocated PID is definitively something for a different ticket/pull, so at least for me this is good. Will merge today if not other concerns. Thanks

**Pull request merged**
